### PR TITLE
Type aliases

### LINKED ISSUE #472: Document type aliases
For the following code:

```python
from typing import Callable, Tuple

Parser = Callable[[str], Tuple[int, bytes, bytes]]
```

Pydoctor generates the documentation:

image

It would be nice if the variable value was documented for variables that are just `typing` aliases.

### PR DESCRIPTION
**Pull request by tristanlatr**
Fixes #472 

Marked as draft until #576 is merged.

### CODE REVIEW

File: pydoctor/astbuilder.py [hunk 1]
```diff
@@ -1322,3 +1302,7 @@ def parseDocformat(node: ast.Assign, mod: model.Module) -> None:
     '__all__': parseAll,
     '__docformat__': parseDocformat
 }
+
+
+def setup_pydoctor_extension(r:extensions.ExtRegistrar) -> None:
+    r.register_astbuilder_visitors(TypeAliasVisitorExt)
```
Comment (tristanlatr) [inline pydoctor/astbuilder.py:1308]:
```suggestion
    r.register_astbuilder_visitor(TypeAliasVisitorExt)
```

File: pydoctor/astutils.py [hunk 2]
```diff
@@ -155,3 +174,167 @@ def is_using_annotations(expr: Optional[ast.AST],
...
+        module = ctx.module
+        assert module is not None
+        module.report(f'syntax error in annotation: {ex}', lineno_offset=node.lineno)
+        return node
+    else:
+        assert isinstance(expr, ast.expr), expr
+        return expr
+
+class _AnnotationStringParser(ast.NodeTransformer):
+    """Implementation of L{ModuleVistor._unstring_annotation()}.
```
Comment (tristanlatr) [inline pydoctor/astutils.py:195]:
```suggestion
    """Implementation of L{unstring_annotation()}.
```

File: pydoctor/astutils.py [hunk 3]
```diff
@@ -155,3 +174,167 @@ def is_using_annotations(expr: Optional[ast.AST],
...
+        "typing.OrderedDict",
+        "typing.Counter",
+        "typing.ChainMap",
+        "typing.Generator",
+        "typing.AsyncGenerator",
+        "typing.Type",
+        "typing.Pattern",
+        "typing.Match",
+        # Special forms
+        "typing.Final",
```
Comment (tristanlatr) [inline pydoctor/astutils.py:287]:
```suggestion
```

File: pydoctor/templatewriter/pages/attributechild.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def functionBody(self, request: object, tag: Tag) -> "Flattenable":
 
     @renderer
     def constantValue(self, request: object, tag: Tag) -> "Flattenable":
-        if self.ob.kind is not DocumentableKind.CONSTANT or self.ob.value is None:
+        if self.ob.kind not in (DocumentableKind.CONSTANT, DocumentableKind.TYPING_VAR) or self.ob.value is None:
```
Comment (tristanlatr) [inline pydoctor/templatewriter/pages/attributechild.py:74]: Maybe we should have a system method determining which object we should show the value for, this would have the benefit to be hackable.

```suggestion
        if not self.ob.sytem._should_present_constant_value_of(self.ob) or self.ob.value is None:
```

File: pydoctor/model.py [hunk 5]
```diff
@@ -693,6 +694,8 @@ def __init__(self, options: Optional['Options'] = None):
             self.extensions = list(extensions.get_extensions())
         assert isinstance(self.extensions, list)
         assert isinstance(self.custom_extensions, list)
+        # pydoctor.astbuilder includes some required extensions, so always add it.
```
Comment (tristanlatr) [inline pydoctor/model.py:697]: This design allows us to separate the logic for new features that only depends on the current ast builder flow to be separated in different classes.

File: pydoctor/test/test_astbuilder.py [hunk 6]
```diff
@@ -1406,7 +1406,7 @@ def test_literal_string_annotation(annotation: str, expected: str) -> None:
     """Strings inside Literal annotations must not be recursively parsed."""
     stmt, = ast.parse(annotation).body
     assert isinstance(stmt, ast.Expr)
-    unstringed = astbuilder._AnnotationStringParser().visit(stmt.value)
+    unstringed = astutils._AnnotationStringParser().visit(stmt.value)
```
Comment (tristanlatr) [inline pydoctor/test/test_astbuilder.py:1409]: Using unstring_annotation() directly would be best I suppose.

**Comment by tristanlatr** [comment]
Hi @not-my-profile, this addresses the issue issue you opened, could you have a look quickly to tell me what you think ? Thanks.

**Comment by not-my-profile** [comment]
I just looked at the two example links you posted, those look good to me :)

**Comment by tristanlatr** [comment]
Probably, for the sake of correctness, we should have two new kinds. One for type aliases and one for type variables.

**Comment by tristanlatr** [comment]
> Probably, for the sake of correctness, we should have two new kinds. One for type aliases and one for type variables.

I’m not quite sure.. @glyph any opinion about that ?

**Comment by glyph** [comment]
> > Probably, for the sake of correctness, we should have two new kinds. One for type aliases and one for type variables.
> 
> I’m not quite sure.. @glyph any opinion about that ?

I don't really know enough about pydoctor's internals to appreciate the significance of a "kind". If you want to give me a little context about the implications of the distinctionI'm happy to provide an opinion though

**Comment by tristanlatr** [comment]
Even if undocumented, the concept of "Kinds" is a public concept of pydoctor, below you can see a screen shot with a couples of different kinds 

by default 2022-06-04 at 9 26 07 PM

You can find the enumeration here: https://pydoctor.readthedocs.io/en/latest/api/pydoctor.model.DocumentableKind.html

In the current PR, I only added one new kind: "Type", and my question was wether we should differentiate type alias and type vars with two kinds or not? 

Tell me what you think,

Thanks,

**Comment by glyph** [comment]
Thanks for the explanation.

I'd say that all of Type, TypeVar, and TypeAlias are distinct kinds of things, and should be called out as such.

For example, `AnyStr` from the stdlib is a `TypeVar`, but the equivalent `Type` would have a completely different meaning.  Consider these two signatures:

```python3
def something(x: AnyStr, y: AnyStr) -> AnyStr:
def something(x: str | bytes, y: str | bytes) -> str | bytes:
```

The first means "takes either str or bytes and returns same" whereas the latter means "takes any combination of str or bytes and returns one of them based on its own criteria"

**Comment by tristanlatr** [comment]
Ok, your probably right, thank. I'll introduce two new kinds: "Type Alias" and "Type Variable".

File: pydoctor/test/test_astbuilder.py [hunk 7]
```diff
@@ -1971,3 +1971,39 @@ class j: pass
...
+        mylst = yourlst = list[str]
+        alist: TypeAlias = 'list[str]'
+        
+        notanalias = 'Callable[[str], Tuple[int, bytes, bytes]]'
+
+        class F:
+            from ext import what
+            L = _j = what.some = list[str]
+            def __init__(self):
+                self.P: TypeAlias = 'Callable[[str], Tuple[int, bytes, bytes]]'
```
[the remark this answers is not in the document]
Comment (glyph) [reply pydoctor/test/test_astbuilder.py:1995]: I agree that this is unlikely to happen, but when it *does* happen, what is the feedback going to be like? At the very least there should be some docs about the limitations here.

But, this is not a blocker: we can do that separately. Having up-to-date docs is great, but the current behavior with type aliases is not an improvement over having it mostly work and edge-cases be undocumented.

[the remark this answers is not in the document]
Comment (glyph) [reply pydoctor/test/test_astbuilder.py]: Oh, good point. Well, nevermind then :)

**Review by glyph — APPROVED** [review]
lGTM
